In [2]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import MinMaxScaler, StandardScaler


# Part 1: Load the dataset

In [3]:
# Load the given datasets
catagorical_dataset = pd.read_csv('/workspaces/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_categorical.csv')
numerical_dataset = pd.read_csv('/workspaces/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_numerical.csv')

# Print the data
print("catagorical data \n")
catagorical_dataset

catagorical data 



,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


In [4]:
print("numerical Data \n")
numerical_dataset

numerical Data 



,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: Features are the input variables used to make predictions. In this dataset, features include various health indicators and demographic information that might influence the outcome regarding CKD. The target (or label) is the variable we are trying to predict. In this case, the target variable is "class," which indicates whether the patient has CKD or not (binary classification). Numerical Features: These are features that represent quantities and can be expressed as numbers. They can be further classified into continuous (which can take any value within a range) and discrete (which can only take specific values). Categorical Features: These features represent categories or groups and are often text labels. They can be nominal (no specific order) or ordinal (with a meaningful order).

Numerical Columns (3 examples):
age: Represents the age of the patient in years (integer).
bp: Represents blood pressure measured in mm/Hg (integer).
sc (serum creatinine): Represents serum creatinine levels measured in mg/dl (continuous).

Categorical Columns (3 examples):
al (albumin): Represents the presence or levels of albumin as a categorical variable.
rbc (red blood cells): A binary feature indicating the presence or absence of red blood cells.
dm (diabetes mellitus): A binary feature indicating whether the patient has diabetes mellitus or not.

A theory I personally have is the numerical data might tell us more about the patient themselves while the catagorical data tells us more about the indicators that show kidney disease. This however is not completely accurate and is just a theory. 


# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [5]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
# Drop these duplicate rows

# Check for duplicate rows based on 'Unique Id'
duplicates_numerical = numerical_dataset[numerical_dataset.duplicated(subset='unique_id', keep=False)]

if not duplicates_numerical.empty:
    print(f"Total number of duplicate rows in numerical dataset: {duplicates_numerical.shape[0]}")
    print("Duplicate rows:")
    print(duplicates_numerical)




# Repeat the same for categorical dataset. Print the duplicate rows and drop them



Total number of duplicate rows in numerical dataset: 59
Duplicate rows:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
15      224481  54.0   60.0  125.0   21.0  1.3  137.0  3.4  15.0  46.0   
35      431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
42      441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
53      261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
54      721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
61      872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
78      501321   4.0    NaN   99.0   23.0  0.6  138.0  4.4  12.0  34.0   
79      283427  55.0   80.0  140.0   49.0  0.5  150.0  4.9  15.7  47.0   
81      302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
98      708255   3.0    NaN    NaN   22.

In [6]:
# Drop duplicate rows
numerical_dataset = numerical_dataset.drop_duplicates(subset='unique_id', keep='first')  # keep='first' keeps the first occurrence

# Print the cleaned numerical dataset
print("\nNumerical dataset after dropping duplicates:")
print(numerical_dataset)


Numerical dataset after dropping duplicates:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2 

In [7]:
# Check for duplicate rows based on 'Unique Id'
duplicates_categorical = catagorical_dataset[catagorical_dataset.duplicated(subset='unique_id', keep=False)]

if not duplicates_categorical.empty:
    print(f"Total number of duplicate rows in categorical dataset: {duplicates_categorical.shape[0]}")
    print("Duplicate rows:")
    print(duplicates_categorical)



Total number of duplicate rows in categorical dataset: 59
Duplicate rows:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
10      708255  2.0  0.0    normal    normal  notpresent  notpresent   no   
34      441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
39      240975  0.0  0.0    normal    normal  notpresent  notpresent   no   
51      354223  0.0  0.0    normal    normal  notpresent  notpresent   no   
71      283427  0.0  0.0    normal    normal  notpresent  notpresent   no   
74      302906  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
88      856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
94      501321  1.0  0.0       NaN    normal  notpresent  notpresent   no   
113     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
124     261137  0.0  0.0    normal    normal  notpresent  notpresent   no   
12

In [8]:
# Drop duplicate rows
catagorical_dataset = catagorical_dataset.drop_duplicates(subset='unique_id', keep='first')  # keep='first' keeps the first occurrence
# Print the cleaned categorical dataset
print("\nCategorical dataset after dropping duplicates:")
print(catagorical_dataset)


Categorical dataset after dropping duplicates:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
396     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
397     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
398     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
399     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
400     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad appet   pe  an

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answe:The numerical dataset had 389 entries while the catagorical had 399 entries. I think all th unique id's are present, since the combined dataset has 399 entries, and there may be some entries that were only present in the catagorical set. 

In [9]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

#Print the combined dataset

combined_df = pd.merge(numerical_dataset, catagorical_dataset, on='unique_id', how='outer')

# Print the combined dataset
print("Combined dataset:")
print(combined_df)

Combined dataset:
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc        

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [10]:
# Calculate the percentage of rows that contain atleast one missing value

# Print %

# Drop these rows from the dataset

# Print the Dataset

# Calculate the percentage of rows with at least one missing value
missing_rows = combined_df.isnull().any(axis=1)  # Get a boolean Series where True indicates a row with at least one NaN
percentage_missing = (missing_rows.sum() / len(combined_df)) * 100  # Calculate percentage

# Print the percentage
print(f"Percentage of rows that contain at least one missing value: {percentage_missing:.2f}%")

# Drop the rows with any missing values
cleaned_df = combined_df.dropna()

# Print the cleaned dataset
print("\nCleaned dataset (after dropping rows with missing values):")
print(cleaned_df)

Percentage of rows that contain at least one missing value: 61.75%

Cleaned dataset (after dropping rows with missing values):
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396 

## Part 3.4: Sort the dataset according to the Labels

In [11]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

# Print the dataset
sorted_df = cleaned_df.sort_values(by='Target')

# Reset the index after sorting
sorted_df = sorted_df.reset_index(drop=True)

# Print the sorted dataset
print("Sorted dataset based on 'Target' column:")
print(sorted_df)

Sorted dataset based on 'Target' column:
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

          

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [12]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# Print the dataset
dummy_df = pd.get_dummies(sorted_df, columns=categorical_columns, prefix=categorical_columns, drop_first=False).astype(int)

# Print the transformed DataFrame
print("Transformed dataset with dummy columns:")
print(dummy_df)

#print(dummy_df.iloc[0])


Transformed dataset with dummy columns:
     unique_id  age  bp  bgr   bu  sc  sod  pot  hemo  pcv  ...  cad_no  \
0       102677   59  70   76  186  15  135    7     7   22  ...       1   
1       337834   48  70  117   56   3  111    2    11   32  ...       1   
2       343710   61  80  173  148   3  135    5     7   24  ...       0   
3       349892   59  70  424   55   1  138    4    12   37  ...       0   
4       353025   40  70  253  150  11  132    5    10   31  ...       1   
..         ...  ...  ..  ...  ...  ..  ...  ...   ...  ...  ...     ...   
148     486397   20  70  123   44   1  135    3    14   44  ...       1   
149     203466   22  60   97   18   1  138    4    13   42  ...       1   
150     482105   44  60   95   46   0  138    4    15   50  ...       1   
151     228849   58  80  100   50   1  140    3    14   50  ...       1   
152     996076   57  60  105   49   1  150    4    15   44  ...       1   

     cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [13]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Print the dataset
# Function to filter out outliers
def remove_outliers(df, numerical_columns):
    for column in numerical_columns:
        mean = df[column].mean()
        std_dev = df[column].std()
        upper_bound = mean + 3 * std_dev
        lower_bound = mean - 3 * std_dev
        df = df[(df[column] <= upper_bound) & (df[column] >= lower_bound)]
    return df

# Remove outliers from the DataFrame
cleaned_df = remove_outliers(dummy_df, numerical_columns)
cleaned_df = cleaned_df.reset_index(drop=True)


print(cleaned_df)

     unique_id  age   bp  bgr   bu  sc  sod  pot  hemo  pcv  ...  cad_no  \
0       343710   61   80  173  148   3  135    5     7   24  ...       0   
1       397388   63  100  280   35   3  143    3    13   40  ...       0   
2       438182   73   80  253  142   4  138    5    10   33  ...       0   
3       474407   21   90  107   40   1  125    3     8   23  ...       1   
4       475200   64  100  163   54   7  140    4     7   26  ...       1   
..         ...  ...  ...  ...  ...  ..  ...  ...   ...  ...  ...     ...   
128     486397   20   70  123   44   1  135    3    14   44  ...       1   
129     203466   22   60   97   18   1  138    4    13   42  ...       1   
130     482105   44   60   95   46   0  138    4    15   50  ...       1   
131     228849   58   80  100   50   1  140    3    14   50  ...       1   
132     996076   57   60  105   49   1  150    4    15   44  ...       1   

     cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0          1   

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [14]:
# Normalize the all Numerical Attributes in the dataset.

# Print the dataset
min_max_scaler = MinMaxScaler()
cleaned_df[min_max_scaler.feature_names_in_] = min_max_scaler.fit_transform(cleaned_df)

print("Min-Max Scaled Data:")
print(cleaned_df)

Min-Max Scaled Data:
     unique_id       age    bp       bgr        bu     sc       sod       pot  \
0     0.264544  0.743243  0.50  0.442060  0.901961  0.375  0.500000  1.000000   
1     0.325059  0.770270  1.00  0.901288  0.163399  0.375  0.766667  0.333333   
2     0.371049  0.905405  0.50  0.785408  0.862745  0.500  0.600000  1.000000   
3     0.411888  0.202703  0.75  0.158798  0.196078  0.125  0.166667  0.333333   
4     0.412782  0.783784  1.00  0.399142  0.287582  0.875  0.666667  0.666667   
..         ...       ...   ...       ...       ...    ...       ...       ...   
128   0.425405  0.189189  0.25  0.227468  0.222222  0.125  0.500000  0.333333   
129   0.106438  0.216216  0.00  0.115880  0.052288  0.125  0.600000  0.666667   
130   0.420566  0.513514  0.00  0.107296  0.235294  0.000  0.600000  0.666667   
131   0.135054  0.702703  0.50  0.128755  0.261438  0.125  0.666667  0.333333   
132   1.000000  0.689189  0.00  0.150215  0.254902  0.125  1.000000  0.666667   

     h

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: I removed only the unique id because that is the only column where none of the information matches the data, it is simply an arbitrary identifier. 

In [15]:
print(cleaned_df.iloc[0])

unique_id         0.264544
age               0.743243
bp                0.500000
bgr               0.442060
bu                0.901961
sc                0.375000
sod               0.500000
pot               1.000000
hemo              0.000000
pcv               0.032258
wbcc              0.395161
rbcc              0.000000
al                0.500000
su                0.000000
rbc_abnormal      1.000000
rbc_normal        0.000000
pc_abnormal       1.000000
pc_normal         0.000000
pcc_notpresent    1.000000
pcc_present       0.000000
ba_notpresent     1.000000
ba_present        0.000000
htn_no            0.000000
htn_yes           1.000000
dm_no             0.000000
dm_yes            1.000000
cad_no            0.000000
cad_yes           1.000000
appet_good        0.000000
appet_poor        1.000000
pe_no             0.000000
pe_yes            1.000000
ane_no            0.000000
ane_yes           1.000000
Target_ckd        1.000000
Target_notckd     0.000000
Name: 0, dtype: float64


In [16]:
#Remove that column
df = cleaned_df.drop(['unique_id'], axis=1)

# Print the dataset
print(df)

          age    bp       bgr        bu     sc       sod       pot  hemo  \
0    0.743243  0.50  0.442060  0.901961  0.375  0.500000  1.000000   0.0   
1    0.770270  1.00  0.901288  0.163399  0.375  0.766667  0.333333   0.6   
2    0.905405  0.50  0.785408  0.862745  0.500  0.600000  1.000000   0.3   
3    0.202703  0.75  0.158798  0.196078  0.125  0.166667  0.333333   0.1   
4    0.783784  1.00  0.399142  0.287582  0.875  0.666667  0.666667   0.0   
..        ...   ...       ...       ...    ...       ...       ...   ...   
128  0.189189  0.25  0.227468  0.222222  0.125  0.500000  0.333333   0.7   
129  0.216216  0.00  0.115880  0.052288  0.125  0.600000  0.666667   0.6   
130  0.513514  0.00  0.107296  0.235294  0.000  0.600000  0.666667   0.8   
131  0.702703  0.50  0.128755  0.261438  0.125  0.666667  0.333333   0.7   
132  0.689189  0.00  0.150215  0.254902  0.125  1.000000  0.666667   0.8   

          pcv      wbcc  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [17]:
# Export the dataframe to a new csv file
df.to_csv('data.csv', index=False) 

# Export the dataframe to a new json file
df.to_json('data.json', orient='records', lines=True)


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [21]:
print(cleaned_df.head(16))


    unique_id       age    bp       bgr        bu     sc       sod       pot  \
0    0.264544  0.743243  0.50  0.442060  0.901961  0.375  0.500000  1.000000   
1    0.325059  0.770270  1.00  0.901288  0.163399  0.375  0.766667  0.333333   
2    0.371049  0.905405  0.50  0.785408  0.862745  0.500  0.600000  1.000000   
3    0.411888  0.202703  0.75  0.158798  0.196078  0.125  0.166667  0.333333   
4    0.412782  0.783784  1.00  0.399142  0.287582  0.875  0.666667  0.666667   
5    0.419651  0.729730  0.00  0.935622  0.169935  0.125  0.333333  0.333333   
6    0.457337  0.675676  0.75  0.253219  0.633987  0.750  0.366667  0.666667   
7    0.998986  0.851351  0.25  0.618026  0.562092  0.750  0.000000  0.333333   
8    0.492853  0.540541  0.00  0.399142  0.535948  0.375  0.700000  0.666667   
9    0.982205  0.756757  0.25  0.223176  0.209150  0.125  0.533333  0.666667   
10   0.954628  0.662162  0.50  0.618026  0.411765  0.375  0.566667  0.666667   
11   0.918308  0.783784  0.00  0.725322 

### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

| unique_id |      age |   bp |      bgr |        bu |    sc |       sod |       pot |  hemo |       pcv |   cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----------|----------|------|----------|-----------|-------|-----------|-----------|-------|-----------|----------|---------|------------|-------------|-------|--------|--------|---------|-------------|----------------|
| 0.264544  | 0.743243 | 0.50 | 0.442060 | 0.901961  | 0.375 | 0.500000  | 1.000000  |   0.0 |  0.032258 |       0.0 |     1.0 |        0.0 |         1.0 |   0.0 |    1.0 |    0.0 |      1.0 |         1.0 |            0.0 |
| 0.325059  | 0.770270 | 1.00 | 0.901288 | 0.163399  | 0.375 | 0.766667  | 0.333333  |   0.6 |  0.548387 |       0.0 |     1.0 |        1.0 |         0.0 |   1.0 |    0.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.371049  | 0.905405 | 0.50 | 0.785408 | 0.862745  | 0.500 | 0.600000  | 1.000000  |   0.3 |  0.322581 |       0.0 |     1.0 |        1.0 |         0.0 |   1.0 |    0.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.411888  | 0.202703 | 0.75 | 0.158798 | 0.196078  | 0.125 | 0.166667  | 0.333333  |   0.1 |  0.000000 |       1.0 |     0.0 |        1.0 |         0.0 |   1.0 |    0.0 |    0.0 |      1.0 |         1.0 |            0.0 |
| 0.412782  | 0.783784 | 1.00 | 0.399142 | 0.287582  | 0.875 | 0.666667  | 0.666667  |   0.0 |  0.096774 |       1.0 |     0.0 |        1.0 |         0.0 |   0.0 |    1.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.419651  | 0.729730 | 0.00 | 0.935622 | 0.169935  | 0.125 | 0.333333  | 0.333333  |   0.0 |  0.064516 |       1.0 |     0.0 |        0.0 |         1.0 |   1.0 |    0.0 |    0.0 |      1.0 |         1.0 |            0.0 |
| 0.457337  | 0.675676 | 0.75 | 0.253219 | 0.633987  | 0.750 | 0.366667  | 0.666667  |   0.2 |  0.193548 |       1.0 |     0.0 |        1.0 |         0.0 |   1.0 |    0.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.998986  | 0.851351 | 0.25 | 0.618026 | 0.562092  | 0.750 | 0.000000  | 0.333333  |   0.2 |  0.161290 |       0.0 |     1.0 |        1.0 |         0.0 |   0.0 |    1.0 |    0.0 |      1.0 |         1.0 |            0.0 |
| 0.492853  | 0.540541 | 0.00 | 0.399142 | 0.535948  | 0.375 | 0.700000  | 0.666667  |   0.2 |  0.161290 |       1.0 |     0.0 |        1.0 |         0.0 |   1.0 |    0.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.982205  | 0.756757 | 0.25 | 0.223176 | 0.209150  | 0.125 | 0.533333  | 0.666667  |   0.5 |  0.516129 |       1.0 |     0.0 |        1.0 |         0.0 |   1.0 |    0.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.954628  | 0.662162 | 0.50 | 0.618026 | 0.411765  | 0.375 | 0.566667  | 0.666667  |   0.3 |  0.354839 |       1.0 |     0.0 |        1.0 |         0.0 |   0.0 |    1.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.918308  | 0.783784 | 0.00 | 0.725322 | 0.313725  | 0.500 | 0.566667  | 1.000000  |   0.2 |  0.193548 |       1.0 |     0.0 |        0.0 |         1.0 |   0.0 |    1.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.827835  | 0.878378 | 0.00 | 0.206009 | 0.751634  | 0.625 | 0.533333  | 0.666667  |   0.4 |  0.387097 |       1.0 |     0.0 |        0.0 |         1.0 |   0.0 |    1.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.811184  | 0.878378 | 0.25 | 0.639485 | 0.470588  | 0.375 | 0.433333  | 0.666667  |   0.3 |  0.322581 |       0.0 |     1.0 |        1.0 |         0.0 |   1.0 |    0.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.577526  | 0.716216 | 0.50 | 1.000000 | 0.163399  | 0.125 | 0.066667  | 0.333333  |   0.3 |  0.387097 |       1.0 |     0.0 |        0.0 |         1.0 |   1.0 |    0.0 |    1.0 |      0.0 |         1.0 |            0.0 |
| 0.571173  | 0.000000 | 0.00 | 0.103004 | 0.372549  | 0.125 | 0.500000  | 0.666667  |   0.2 |  0.225806 |       1.0 |     0.0 |        0.0 |         1.0 |   1.0 |    0.0 |    1.0 |      0.0 |         1.0 |            0.0 |


## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [22]:
# Code to rename all the columns in the dataset
import pandas as pd

# Creating a sample DataFrame. Replace this with your actual DataFrame.
data = {
    "unique_id": [0.264544, 0.325059, 0.371049],
    "age": [0.743243, 0.770270, 0.905405],
    "bp": [0.50, 1.00, 0.50],
    "bgr": [0.442060, 0.901288, 0.785408],
    "bu": [0.901961, 0.163399, 0.862745],
    "sc": [0.375, 0.375, 0.500],
    "sod": [0.500000, 0.766667, 0.600000],
    "pot": [1.000000, 0.333333, 1.000000],
    "hemo": [0.0, 0.6, 0.3],
    "pcv": [0.032258, 0.548387, 0.322581],
    "cad_no": [0.0, 0.0, 0.0],
    "cad_yes": [1.0, 1.0, 1.0],
    "appet_good": [0.0, 1.0, 1.0],
    "appet_poor": [1.0, 0.0, 0.0],
    "pe_no": [0.0, 1.0, 1.0],
    "pe_yes": [1.0, 0.0, 0.0],
    "ane_no": [0.0, 1.0, 1.0],
    "ane_yes": [1.0, 0.0, 0.0],
    "Target_ckd": [1.0, 1.0, 1.0],
    "Target_notckd": [0.0, 0.0, 0.0]
}

df = pd.DataFrame(data)

# Define the mapping
rename_dict = {
    "unique_id": "Unique ID",
    "age": "Age",
    "bp": "Blood Pressure",
    "bgr": "Blood Glucose Random",
    "bu": "Blood Urea",
    "sc": "Serum Creatinine",
    "sod": "Sodium",
    "pot": "Potassium",
    "hemo": "Hemoglobin",
    "pcv": "Packed Cell Volume",
    "cad_no": "Coronary Artery Disease (No)",
    "cad_yes": "Coronary Artery Disease (Yes)",
    "appet_good": "Appetite (Good)",
    "appet_poor": "Appetite (Poor)",
    "pe_no": "Pedal Edema (No)",
    "pe_yes": "Pedal Edema (Yes)",
    "ane_no": "Anemia (No)",
    "ane_yes": "Anemia (Yes)",
    "Target_ckd": "Target (Chronic Kidney Disease)",
    "Target_notckd": "Target (Not Chronic Kidney Disease)"
}

# Rename the columns
df.rename(columns=rename_dict, inplace=True)

# Display the updated DataFrame
print(df)


   Unique ID       Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0   0.264544  0.743243             0.5              0.442060    0.901961   
1   0.325059  0.770270             1.0              0.901288    0.163399   
2   0.371049  0.905405             0.5              0.785408    0.862745   

   Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed Cell Volume  \
0             0.375  0.500000   1.000000         0.0            0.032258   
1             0.375  0.766667   0.333333         0.6            0.548387   
2             0.500  0.600000   1.000000         0.3            0.322581   

   Coronary Artery Disease (No)  Coronary Artery Disease (Yes)  \
0                           0.0                            1.0   
1                           0.0                            1.0   
2                           0.0                            1.0   

   Appetite (Good)  Appetite (Poor)  Pedal Edema (No)  Pedal Edema (Yes)  \
0              0.0              1.0              

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [23]:
# Converted SQL to Pandas code

# Sample DataFrame. Replace this with your actual DataFrame.
data = {
    "Unique ID": [0.264544, 0.325059, 0.371049, 0.264544, 0.264544],
    "Age": [0.743243, 0.770270, 0.905405, 0.743243, 0.371049],
    "Target": ["ckd", "notckd", "ckd", "notckd", "ckd"]
}

df = pd.DataFrame(data)

# Group by 'Target' and count the number of occurrences
result = df.groupby('Target').size().reset_index(name='count')

# Display the resulting DataFrame
print(result)


   Target  count
0     ckd      3
1  notckd      2


*Describe what the above code does here*
What This Code Does
Grouping: The query is grouping the DataFrame `df` by the values in the `Target` column. This means it consolidates the data based on unique values in `Target`.
Counting: For each unique value in the `Target` column, it counts the total number of occurrences.
Resulting DataFrame: The resulting DataFrame `result` will contain two columns:
- `Target`: Each unique value from the `Target` column of the original DataFrame.
- `count`: The number of times each unique value appears in the original DataFrame.